### DISCUSSION TITLE
Removal of validateOnly test

### DISCUSSION THREAD

**Discussion by acurrieclark** [livewire/livewire#4333] [7. Bugs (archived)]
Hello all. I have been keeping a track of any changes to the `validateOnly` function after I submitted #3596 in fear that I had broken something for someone else, and it appears I was correct!

Apologies for the issue brought up in #4278, I should have thought that common prefixes apply even when wildcards aren't in use. I am grateful to be put right.

My query regards [the test](https://github.com/livewire/livewire/commit/38abc128446136ba03caa555ed1eec7e013cbe8f#diff-66d64c9d1ac4fc7cc9910a37d735e28d93425b0497acd979b168f5f5017f461eL200) removed in #4319. Is there another way that this validation can be achieved?

For example, let's say I have a component with the validation rules:

```php
protected $rules = [
    'notes.*' => 'max:20',
    'notes.*.title' => 'required',
    'notes.*.content' => 'required|min:10',
    'comments.*.user' => 'required|int',
    'notes' => 'string|max:500|nullable',
];
```
and you wanted to run validations on *all of the notes*, what validation command would you run? 

It seems to me that if you called `$this->validateOnly('notes.*');` you would be looking to validate all of your notes, including finding errors on `notes.*.title` on each one?

**Comment by joshhanley** [comment]
@acurrieclark actually the break in #4278 was actually my fault haha! 😆 But yeah there were some issues before that, but it's all good, I think we've now sorted them and I think it's a great feature! 🙂

Regarding the test that was removed, I reverted back to 2.5.5 to see what the behaviour was then, and at that point validateOnly would only use the single rule of `'notes.*' => 'max:20'` but none of the others.

As such, due to the issues we'd been having with the feature, I thought it best to ensure it was at least running as it was before.

Ideally, that change to support cascading rules like you are describing above should have been put into a seperate PR, as we didn't actually realise that was a change you had made.

To me that is a different change than getting validation working for a specific collection item like using `notes.*.title` rule to validate specifically only say `notes.1.title` by calling `->validateOnly('notes.1.title');` which I believe was the original intent yeah?

So in regards to how to achieve the cascading rules like above, to say validate a collection and all of its properties, my first question would be, how does Laravel handle something like that? Now that's not to say we can't add our own custom stuff, but it's always good to check to see how someone would expect to do something like that with Laravel first.

Personally I'm not sure if I prefer `validateOnly` being specific (like it is now) or to cascade the rules like you are suggesting.

On one hand I can see it being useful to call say `->validateOnly('notes.1')` and all the properties for note 1 get validated, but then the rule of `'notes.*' => 'max:20'` should actually be ignored, as we're not validating the collection, but a single instance, especially if the rule was actually something like `'notes.*' => 'min:5'`.

On the other hand though, when we call `->validateOnly('notes.*')`, maybe the user wants to only validate that the collection has a max of 20 items and isn't expecting all the properties of the notes in the collection to be validated too.

What do you think?

**Comment by acurrieclark** [comment]
Hey Josh, thanks for getting back to me.

I see what you mean when you say that it should have been a separate pull request. I hadn't really thought of it as a separate issue, and when I added the test it was more because I thought the test was missing than because I had changed anything! I appreciate the feedback though. A lesson learned.

The interesting thing about this topic is that Laravel doesn't appear to have a `validateOnly` method, so all we can do is look at the main `Illuminate\Support\Facades\Validator`.

It seems that in Laravel, the behaviour is that `notes` would be used to validate the array (array size, type, required etc) and `notes.*` would be used to validate the array values.

For example:
```php
$data = [
    "notes" => ['a']
];

Validator::make($data, [
    "notes"    => "required|array|min:3",
    "notes.*"  => "required|string|distinct|min:5",
])->errors()->all();

// results in the errors:

[
     "The notes must have at least 3 items.",
     "The notes.0 must be at least 5 characters.",
]
```

The first error comes from the array not having enough entries, and the last error from the existing entry not having enough characters. So `notes.*` is actually validating the array values, rather than the array itself.

However, if a `note` also has child elements, it is less clear. For example:

```php
$data = [
    "notes" => [["title" => 'a'], 'b', 'c']
];

Validator::make($data, [
    "notes"    => "required|array|min:3",
    "notes.*"  => "required|array",
    "notes.*.title" => "required"
])->errors()->all();

// results in

[
     "The notes.1 must be an array.",
     "The notes.2 must be an array.",
     "The notes.1.title field is required.",
     "The notes.2.title field is required.",
]
```
where the first 2 errors are as a result of `notes.*` and the last 2 as a result of `notes.*.title`.

Perhaps `notes.*.*` would be appropriate to test all of the notes? But then, let's say we were to have a set of validation rules:

```php
[
    "notes"    => "required|array|min:3",
    "notes.*.title" => "required"
]
```
with no mention of `notes.*` I would imagine it would be confusing to a user as to why `notes.*` didn't validate all of their notes.

All of that said, the key use for `validateOnly` is obviously for validating the input which has just been updated, usually via the `updated` method. In that instance, the parameter passed in is never going to include a wildcard anyway.

I am not really sure what to think!

**Reply by joshhanley** [reply]
@acurrieclark no worries! All good 🙂

Ah you are right about Laravel having nothing like that.

Yeah with Laravel you are not specifying which property to validate instead you are just telling it what rules to run against the request data.

So my thoughts would be that technically Livewire is the same in the sense that you should pass in which rules to use, to run validation against component data, rather than specifying which property should be validated (and any sub data within it).

I think a better solution might be to actually allow an array of rules to be passed into `validateOnly` that way you don't need to specify all of the rules and their values again, but just "select" or filter from the ones that are available in the `$rules` property.

I could also imagine creating another validation method like `validateProperty` which then works the way you are talking about with the cascading rules.

I'm not sure which would be better/ more helpful.

It might be worth opening up a feature ideas discussion (and marking this bug report answered) and outlining the possibilities so others can weigh in on them.

Would you be able to do that?
